app/src/main/java/org/onionshare/android/files/FileManager.kt

```diff
@@ -86,55 +78,47 @@ class FileManager @Inject constructor(
             }
             SendFile(name, sizeHuman, size, uri, documentFile.type)
         }
-        _state.value = Added(existingFiles + files)
+        _filesState.value = FilesState(existingFiles + files)
     }
 
     fun removeFile(file: SendFile) {
-        checkModificationIsAllowed()
-
         // release persistable Uri permission again
         file.releaseUriPermission()
 
-        val newList = state.value.files.filterNot { it == file }
-        _state.value = Added(newList)
+        val newList = filesState.value.files.filterNot { it == file }
+        _filesState.value = FilesState(newList)
     }
 
     fun removeAll() {
-        checkModificationIsAllowed()
-
         // release persistable Uri permissions again
-        state.value.files.iterator().forEach { file ->
+        filesState.value.files.iterator().forEach { file ->
             file.releaseUriPermission()
         }
-        _state.value = Added(emptyList())
+        _filesState.value = FilesState(emptyList())
     }
 
-    suspend fun zipFiles() = withContext(Dispatchers.IO) {
-        if (state.value is Zipped) return@withContext
-        zipJob?.cancelAndJoin()
-        zipJob = launch {
-            try {
-                zipFilesInternal()
-            } catch (e: FileErrorException) {
-                // remove errorFile from list of files, so user can try again
-                val newFiles = state.value.files.toMutableList().apply { remove(e.file) }
-                _state.value = FilesState.Error(newFiles, e.file)
-            } catch (e: IOException) {
-                _state.value = FilesState.Error(state.value.files)
-            }
+    suspend fun zipFiles(): ZipResult = withContext(Dispatchers.IO) {
+        try {
+            val sendPage = zipFilesInternal()
+            ZipResult.Zipped(sendPage)
+        } catch (e: FileErrorException) {
+            // remove errorFile from list of files, so user can try again
+            val newFiles = filesState.value.files.toMutableList().apply { remove(e.file) }
+            _filesState.value = FilesState(newFiles)
+            ZipResult.Error(e.file)
+        } catch (e: Exception) {
+            ZipResult.Error()
```

Comment:
Looks to me like a CancellationException from `zipFilesInternal()` would be caught here, maybe resulting in the error UI being shown even if the cancellation was triggered by the user - am I reading that right?

Reply:
Good point. I tried provoking this, but the cancellation seems to prevent this state from being combined into a new ShareUiState. If not, we'd need some third ZipResult state like cancelled which would be nice to avoid.

Reply:
On second reading, I think it would be possible to run into a problem here, but please check my thinking:

1. `startSharingJob` starts a `fileTask` (which calls the method we're looking at) and a `torTask`
2. It waits for the `torTask` to finish, which means the Tor process is running or there was an error
3. It starts an `observerTask` to monitor the progress of the zipping and Tor bootstrapping (if Tor failed to start, I guess the observer notices?)
4. It checks that `startSharingJob` hasn't been cancelled yet
5. It waits for the `fileTask` to finish, which means the file have been zipped or there was an error
6. If the files have been zipped, it continues with startup. Else it emits `ShareUiState.ErrorAddingFile`

If this reading is right, then if the user cancels startup at item 5, the `fileTask` will throw a `CancellationException` resulting in `ShareUiState.ErrorAddingFile`. To reproduce this we'd need to zip a large amount of data, so that zipping took longer than starting the Tor process and there was a reasonably long wait between the check for cancellation at item 4 and the zipping being complete, where we'd have a chance to hit the cancel button and cause the exception.

Reply:
I had tested this with a 1GB file last time. There, I think `zipFilesInternal()` was catching the `CancellationException` when after it finished zipping the file (we still don't do `ensureActive()` while zipping a while).

 Ideally, we can write tests for those scenarios as well. Hope I'll find the time.

Reply:
While trying to test this with a large file I ran into something strange.

It seems that while the files are being zipped, our handling of the Tor state updates from onionwrapper doesn't work as expected. Tor successfully bootstraps, and yet we fail to notice this and eventually conclude that it has failed to start.

If I try it with a small file, everything's fine. The file gets zipped, Tor boostraps and we detect that it's bootstrapped, we publish the hidden service and the UI moves to the sharing state.

But if I choose a file that's large enough for zipping to take longer than bootstrapping, the weirdness happens. I can reproduce this on the Samsung A21s with a file containing 1 GB of random data, which takes about 3 minutes to zip. Here's the log:

```
2023-07-21 15:51:58.546 27874-28033 o.o.a.t.TorManager      org.onionshare.android.nightly       I  Starting...
2023-07-21 15:51:58.554 27874-28033 TorWrapper              org.onionshare.android.nightly       I  Starting Tor
2023-07-21 15:51:58.582 27874-28019 TorWrapper              org.onionshare.android.nightly       I  Jul 21 15:51:58.581 [notice] Tor 0.4.7.13 (git-7c1601fb6edd780f) running on Linux with Libevent 2.1.12-stable, OpenSSL 1.1.1t, Zlib 1.2.11, Liblzma 5.2.10, Libzstd 1.4.9 and Unknown N/A as libc.
2023-07-21 15:51:58.672 27874-27874 o.o.a.t.ShareService    org.onionshare.android.nightly       D  onStartCommand Intent { cmp=org.onionshare.android.nightly/org.onionshare.android.tor.ShareService }
2023-07-21 15:51:58.683 27874-27874 o.o.a.t.TorManager      org.onionshare.android.nightly       I  new state: STARTING
2023-07-21 15:51:58.790 27874-28033 TorWrapper              org.onionshare.android.nightly       I  Auth cookie created
2023-07-21 15:52:00.351 27874-27874 o.o.a.t.TorManager      org.onionshare.android.nightly       I  new state: STARTED
2023-07-21 15:52:00.352 27874-28033 RenewableWakeLock       org.onionshare.android.nightly       I  Acquiring wake lock org.onionshare.android.nightly
2023-07-21 15:52:00.352 27874-27988 o.o.a.t.TorManager      org.onionshare.android.nightly       I  Starting check job
2023-07-21 15:52:00.353 27874-27988 o.o.a.t.TorManager      org.onionshare.android.nightly       I  Waiting for Tor to start...
2023-07-21 15:52:00.356 27874-27874 o.o.a.t.TorManager      org.onionshare.android.nightly       I  new state: CONNECTING
2023-07-21 15:52:00.418 27874-28498 TorWrapper              org.onionshare.android.nightly       I  NOTICE Opening Socks listener on 127.0.0.1:53054
2023-07-21 15:52:00.418 27874-28498 TorWrapper              org.onionshare.android.nightly       I  NOTICE Opened Socks listener connection (ready) on 127.0.0.1:53054
2023-07-21 15:52:00.418 27874-28033 o.o.a.ShareManager      org.onionshare.android.nightly       I  Tor task returned
2023-07-21 15:52:00.419 27874-28033 o.o.a.ShareManager      org.onionshare.android.nightly       I  Awaiting file task...
2023-07-21 15:52:00.419 27874-28032 o.o.a.ShareManager      org.onionshare.android.nightly       I  Starting Observer task...
2023-07-21 15:52:00.422 27874-28031 o.o.a.ShareManager      org.onionshare.android.nightly       I  New share state: Starting(zipPercent=0, torPercent=5)
2023-07-21 15:52:01.373 27874-28498 TorWrapper              org.onionshare.android.nightly       I  OR connection LAUNCHED
2023-07-21 15:52:01.374 27874-28498 TorWrapper              org.onionshare.android.nightly       I  NOTICE Bootstrapped 5% (conn): Connecting to a relay
2023-07-21 15:52:01.377 27874-27990 o.o.a.ShareManager      org.onionshare.android.nightly       I  New share state: Starting(zipPercent=0, torPercent=10)
2023-07-21 15:52:01.501 27874-28498 TorWrapper              org.onionshare.android.nightly       I  NOTICE Bootstrapped 10% (conn_done): Connected to a relay
2023-07-21 15:52:01.504 27874-27990 o.o.a.ShareManager      org.onionshare.android.nightly       I  New share state: Starting(zipPercent=0, torPercent=14)
2023-07-21 15:52:01.656 27874-28498 TorWrapper              org.onionshare.android.nightly       I  NOTICE Bootstrapped 14% (handshake): Handshaking with a relay
2023-07-21 15:52:01.664 27874-27990 o.o.a.ShareManager      org.onionshare.android.nightly       I  New share state: Starting(zipPercent=0, torPercent=18)
2023-07-21 15:52:01.966 27874-28498 TorWrapper              org.onionshare.android.nightly       I  NOTICE Bootstrapped 15% (handshake_done): Handshake with a relay done
2023-07-21 15:52:01.967 27874-28498 TorWrapper              org.onionshare.android.nightly       I  NOTICE Bootstrapped 75% (enough_dirinfo): Loaded enough directory info to build circuits
2023-07-21 15:52:01.968 27874-28498 TorWrapper              org.onionshare.android.nightly       I  NOTICE Bootstrapped 90% (ap_handshake_done): Handshake finished with a relay to build circuits
2023-07-21 15:52:01.968 27874-28498 TorWrapper              org.onionshare.android.nightly       I  NOTICE Bootstrapped 95% (circuit_create): Establishing a Tor circuit
2023-07-21 15:52:01.969 27874-27990 o.o.a.ShareManager      org.onionshare.android.nightly       I  New share state: Starting(zipPercent=0, torPercent=19)
2023-07-21 15:52:01.969 27874-28498 TorWrapper              org.onionshare.android.nightly       I  OR connection CONNECTED
2023-07-21 15:52:01.969 27874-28498 TorWrapper              org.onionshare.android.nightly       I  1 OR connections connected
2023-07-21 15:52:01.970 27874-27990 o.o.a.ShareManager      org.onionshare.android.nightly       I  New share state: Starting(zipPercent=0, torPercent=91)
2023-07-21 15:52:02.402 27874-28498 TorWrapper              org.onionshare.android.nightly       I  OR connection LAUNCHED
2023-07-21 15:52:02.420 27874-28498 TorWrapper              org.onionshare.android.nightly       I  Circuit built
2023-07-21 15:52:02.421 27874-28498 TorWrapper              org.onionshare.android.nightly       I  NOTICE Bootstrapped 100% (done): Done
2023-07-21 15:52:02.422 27874-28498 TorWrapper              org.onionshare.android.nightly       I  Bootstrapped
2023-07-21 15:52:02.430 27874-27874 o.o.a.t.TorManager      org.onionshare.android.nightly       I  new state: CONNECTED
2023-07-21 15:52:02.432 27874-28033 o.o.a.ShareManager      org.onionshare.android.nightly       I  New share state: Starting(zipPercent=0, torPercent=95)
2023-07-21 15:52:02.730 27874-28498 TorWrapper              org.onionshare.android.nightly       I  OR connection CONNECTED
2023-07-21 15:52:02.730 27874-28498 TorWrapper              org.onionshare.android.nightly       I  2 OR connections connected
2023-07-21 15:53:00.355 27874-28028 RenewableWakeLock       org.onionshare.android.nightly       I  Renewing wake lock org.onionshare.android.nightly
2023-07-21 15:54:00.361 27874-28028 RenewableWakeLock       org.onionshare.android.nightly       I  Renewing wake lock org.onionshare.android.nightly
2023-07-21 15:54:02.482 27874-28033 o.o.a.t.TorManager      org.onionshare.android.nightly       I  Tor startup is not making progress
2023-07-21 15:54:02.482 27874-28033 o.o.a.t.TorManager      org.onionshare.android.nightly       I  Getting bridges from Moat...
2023-07-21 15:54:05.910 27874-28033 o.o.a.t.TorManager      org.onionshare.android.nightly       I  Using built-in bridges...
2023-07-21 15:54:05.935 27874-28033 o.o.a.t.TorManager      org.onionshare.android.nightly       I  Using bridges:
2023-07-21 15:54:05.935 27874-28033 o.o.a.t.TorManager      org.onionshare.android.nightly       I    Bridge obfs4 192.95.36.142:443 CDF2E852BF539B82BD10E27E9115A31734E378C2 cert=qUVQ0srL1JI/vO6V6m/24anYXiJD3QP2HgzUKQtQ7GRqqUvs7P+tG43RtAqdhLOALP7DJQ iat-mode=1
2023-07-21 15:54:05.935 27874-28033 o.o.a.t.TorManager      org.onionshare.android.nightly       I    Bridge obfs4 37.218.245.14:38224 D9A82D2F9C2F65A18407B1D2B764F130847F8B5D cert=bjRaMrr1BRiAW8IE9U5z27fQaYgOhX1UCmOpg2pFpoMvo6ZgQMzLsaTzzQNTlm7hNcb+Sg iat-mode=0
2023-07-21 15:54:05.935 27874-28033 o.o.a.t.TorManager      org.onionshare.android.nightly       I    Bridge obfs4 85.31.186.98:443 011F2599C0E9B27EE74B353155E244813763C3E5 cert=ayq0XzCwhpdysn5o0EyDUbmSOx3X/oTEbzDMvczHOdBJKlvIdHHLJGkZARtT4dcBFArPPg iat-mode=0
2023-07-21 15:54:05.935 27874-28033 o.o.a.t.TorManager      org.onionshare.android.nightly       I    Bridge obfs4 85.31.186.26:443 91A6354697E6B02A386312F68D82CF86824D3606 cert=PBwr+S8JTVZo6MPdHnkTwXJPILWADLqfMGoVvhZClMq/Urndyd42BwX9YFJHZnBB3H0XCw iat-mode=0
2023-07-21 15:54:05.936 27874-28033 o.o.a.t.TorManager      org.onionshare.android.nightly       I    Bridge obfs4 193.11.166.194:27015 2D82C2E354D531A68469ADF7F878FA6060C6BACA cert=4TLQPJrTSaDffMK7Nbao6LC7G9OW/NHkUwIdjLSS3KYf0Nv4/nQiiI8dY2TcsQx01NniOg iat-mode=0
2023-07-21 15:54:05.936 27874-28033 o.o.a.t.TorManager      org.onionshare.android.nightly       I    Bridge obfs4 193.11.166.194:27020 86AC7B8D430DAC4117E9F42C9EAED18133863AAF cert=0LDeJH4JzMDtkJJrFphJCiPqKx7loozKN7VNfuukMGfHO0Z8OGdzHVkhVAOfo1mUdv9cMg iat-mode=0
2023-07-21 15:54:05.936 27874-28033 o.o.a.t.TorManager      org.onionshare.android.nightly       I    Bridge obfs4 193.11.166.194:27025 1AE2C08904527FEA90C4C4F8C1083EA59FBC6FAF cert=ItvYZzW5tn6v3G4UnQa6Qz04Npro6e81AP70YujmK/KXwDFPTs3aHXcHp4n8Vt6w/bv8cA iat-mode=0
2023-07-21 15:54:05.936 27874-28033 o.o.a.t.TorManager      org.onionshare.android.nightly       I    Bridge obfs4 209.148.46.65:443 74FAD13168806246602538555B5521A0383A1875 cert=ssH+9rP8dG2NLDN2XuFw63hIO/9MNNinLmxQDpVa+7kTOa9/m+tGWT1SmSYpQ9uTBGa6Hw iat-mode=0
2023-07-21 15:54:05.936 27874-28033 o.o.a.t.TorManager      org.onionshare.android.nightly       I    Bridge obfs4 146.57.248.225:22 10A6CD36A537FCE513A322361547444B393989F0 cert=K1gDtDAIcUfeLqbstggjIw2rtgIKqdIhUlHp82XRqNSq/mtAjp1BIC9vHKJ2FAEpGssTPw iat-mode=0
2023-07-21 15:54:05.936 27874-28033 o.o.a.t.TorManager      org.onionshare.android.nightly       I    Bridge obfs4 45.145.95.6:27015 C5B7CD6946FF10C5B3E89691A7D3F2C122D2117C cert=TD7PbUO0/0k6xYHMPW3vJxICfkMZNdkRrb63Zhl5j9dW3iRGiCx0A7mPhe5T2EDzQ35+Zw iat-mode=0
2023-07-21 15:54:05.936 27874-28033 o.o.a.t.TorManager      org.onionshare.android.nightly       I    Bridge obfs4 51.222.13.177:80 5EDAC3B810E12B01F6FD8050D2FD3E277B289A08 cert=2uplIpLQ0q9+0qMFrK5pkaYRDOe460LL9WHBvatgkuRr/SL31wBOEupaMMJ6koRE6Ld0ew iat-mode=0
2023-07-21 15:54:05.936 27874-28033 o.o.a.t.TorManager      org.onionshare.android.nightly       I    Bridge 92.243.15.235:9001 477EAD3C04036B48235F1F27FC91420A286A4B7F
2023-07-21 15:54:05.936 27874-28033 o.o.a.t.TorManager      org.onionshare.android.nightly       I    Bridge 213.108.108.145:17674 A39C0FE47963B6E8CFE9815549864DE544935A31
2023-07-21 15:54:05.936 27874-28033 o.o.a.t.TorManager      org.onionshare.android.nightly       I    Bridge 213.196.191.96:9060 05E222E2A8C234234FE0CEB58B08A93B8FC360DB
2023-07-21 15:54:05.936 27874-28033 o.o.a.t.TorManager      org.onionshare.android.nightly       I    Bridge 87.100.193.2:9010 13FB63452AADFA082BD2BC3E1E320AD301F07877
2023-07-21 15:54:05.936 27874-28033 o.o.a.t.TorManager      org.onionshare.android.nightly       I    Bridge 65.21.240.163:33245 20BD59649212CFE7412BFC9B94C3CCCFD8F807A8
2023-07-21 15:54:05.936 27874-28033 o.o.a.t.TorManager      org.onionshare.android.nightly       I    Bridge 195.201.202.125:443 D44C0BC5AF900547704BCE5062E4B169672120E8
2023-07-21 15:54:05.937 27874-28033 o.o.a.t.TorManager      org.onionshare.android.nightly       I    Bridge 80.67.179.20:143 D1B6871457A0AA01BF46EEB1989D89086E5E6825
2023-07-21 15:54:05.937 27874-28033 o.o.a.t.TorManager      org.onionshare.android.nightly       I    Bridge 164.163.175.219:9001 42F44429F1F8CD25C9892B845E3FC4374DBB05B1
2023-07-21 15:54:05.937 27874-28033 o.o.a.t.TorManager      org.onionshare.android.nightly       I    Bridge 145.239.252.13:9001 59F622BDF6888CE480038F899CBCC2B7438B19A3
2023-07-21 15:54:05.950 27874-28033 o.o.a.t.TorManager      org.onionshare.android.nightly       I  Waiting for Tor to start...
2023-07-21 15:54:05.950 27874-28498 TorWrapper              org.onionshare.android.nightly       I  NOTICE Switching to guard context "bridges" (was using "default")
2023-07-21 15:54:05.950 27874-28033 o.o.a.t.TorManager      org.onionshare.android.nightly       I  Tor startup is not making progress
2023-07-21 15:54:05.950 27874-28033 o.o.a.t.TorManager      org.onionshare.android.nightly       I  Could not connect to Tor
2023-07-21 15:54:05.951 27874-28033 o.o.a.t.TorManager      org.onionshare.android.nightly       I  Check job finished
2023-07-21 15:54:05.952 27874-27988 o.o.a.ShareManager      org.onionshare.android.nightly       I  New share state: ErrorStarting(torFailedToConnect=true, errorMsg=null)
2023-07-21 15:54:05.954 27874-27990 o.o.a.ShareManager      org.onionshare.android.nightly       I  Stopping sharing...
2023-07-21 15:54:05.954 27874-27990 o.o.a.ShareManager      org.onionshare.android.nightly       I  Wait for start job to finish...
2023-07-21 15:54:06.604 27874-28498 TorWrapper              org.onionshare.android.nightly       I  NOTICE Delaying directory fetches: No running bridges
2023-07-21 15:54:06.605 27874-28498 TorWrapper              org.onionshare.android.nightly       I  OR connection LAUNCHED
2023-07-21 15:54:06.605 27874-28498 TorWrapper              org.onionshare.android.nightly       I  OR connection LAUNCHED
2023-07-21 15:54:06.606 27874-28498 TorWrapper              org.onionshare.android.nightly       I  OR connection LAUNCHED
2023-07-21 15:54:06.607 27874-28498 TorWrapper              org.onionshare.android.nightly       I  OR connection LAUNCHED
2023-07-21 15:54:06.608 27874-28498 TorWrapper              org.onionshare.android.nightly       I  OR connection LAUNCHED
2023-07-21 15:54:06.609 27874-28498 TorWrapper              org.onionshare.android.nightly       I  OR connection LAUNCHED
2023-07-21 15:54:06.609 27874-28498 TorWrapper              org.onionshare.android.nightly       I  OR connection LAUNCHED
2023-07-21 15:54:06.611 27874-28498 TorWrapper              org.onionshare.android.nightly       I  OR connection LAUNCHED
2023-07-21 15:54:06.612 27874-28498 TorWrapper              org.onionshare.android.nightly       I  OR connection LAUNCHED
2023-07-21 15:54:06.612 27874-28498 TorWrapper              org.onionshare.android.nightly       I  OR connection LAUNCHED
2023-07-21 15:54:06.612 27874-28498 TorWrapper              org.onionshare.android.nightly       I  OR connection LAUNCHED
2023-07-21 15:54:06.613 27874-28498 TorWrapper              org.onionshare.android.nightly       I  OR connection LAUNCHED
2023-07-21 15:54:06.613 27874-28498 TorWrapper              org.onionshare.android.nightly       I  OR connection LAUNCHED
2023-07-21 15:54:06.614 27874-28498 TorWrapper              org.onionshare.android.nightly       I  OR connection LAUNCHED
2023-07-21 15:54:06.614 27874-28498 TorWrapper              org.onionshare.android.nightly       I  OR connection LAUNCHED
2023-07-21 15:54:06.614 27874-28498 TorWrapper              org.onionshare.android.nightly       I  OR connection LAUNCHED
2023-07-21 15:54:06.616 27874-28498 TorWrapper              org.onionshare.android.nightly       I  OR connection LAUNCHED
2023-07-21 15:54:06.617 27874-28498 TorWrapper              org.onionshare.android.nightly       I  OR connection LAUNCHED
2023-07-21 15:54:06.618 27874-28498 TorWrapper              org.onionshare.android.nightly       I  OR connection LAUNCHED
2023-07-21 15:54:06.618 27874-28498 TorWrapper              org.onionshare.android.nightly       I  OR connection LAUNCHED
2023-07-21 15:54:06.647 27874-28498 TorWrapper              org.onionshare.android.nightly       I  OR connection CONNECTED
2023-07-21 15:54:06.648 27874-28498 TorWrapper              org.onionshare.android.nightly       I  3 OR connections connected
2023-07-21 15:54:06.674 27874-28498 TorWrapper              org.onionshare.android.nightly       I  OR connection CONNECTED
2023-07-21 15:54:06.674 27874-28498 TorWrapper              org.onionshare.android.nightly       I  4 OR connections connected
2023-07-21 15:54:06.677 27874-28498 TorWrapper              org.onionshare.android.nightly       I  OR connection CONNECTED
2023-07-21 15:54:06.678 27874-28498 TorWrapper              org.onionshare.android.nightly       I  5 OR connections connected
2023-07-21 15:54:06.815 27874-28498 TorWrapper              org.onionshare.android.nightly       I  NOTICE Bridge 'Unnamed' has both an IPv4 and an IPv6 address.  Will prefer using its IPv4 address (145.239.252.13:9001) based on the configured Bridge address.
2023-07-21 15:54:06.815 27874-28498 TorWrapper              org.onionshare.android.nightly       I  NOTICE new bridge descriptor 'Unnamed' (fresh): $59F622BDF6888CE480038F899CBCC2B7438B19A3~Unnamed [f/xDwIe2Xke2hD200rGHtvRt4hsn24A8qjhFW+SyyBM] at 145.239.252.13 and [2001:41d0:800:60d::ff]
2023-07-21 15:54:06.816 27874-28498 TorWrapper              org.onionshare.android.nightly       I  NOTICE Our directory information is no longer up-to-date enough to build circuits: We're missing descriptors for 3/3 of our primary entry guards (total microdescriptors: 7087/7087). That's ok. We will try to fetch missing descriptors soon.
2023-07-21 15:54:06.817 27874-28498 TorWrapper              org.onionshare.android.nightly       I  OR connection CONNECTED
2023-07-21 15:54:06.817 27874-28498 TorWrapper              org.onionshare.android.nightly       I  6 OR connections connected
2023-07-21 15:54:06.817 27874-28498 TorWrapper              org.onionshare.android.nightly       I  OR connection CONNECTED
2023-07-21 15:54:06.817 27874-28498 TorWrapper              org.onionshare.android.nightly       I  7 OR connections connected
2023-07-21 15:54:06.817 27874-28498 TorWrapper              org.onionshare.android.nightly       I  OR connection CONNECTED
2023-07-21 15:54:06.817 27874-28498 TorWrapper              org.onionshare.android.nightly       I  8 OR connections connected
2023-07-21 15:54:06.822 27874-28498 TorWrapper              org.onionshare.android.nightly       I  OR connection CONNECTED
2023-07-21 15:54:06.822 27874-28498 TorWrapper              org.onionshare.android.nightly       I  9 OR connections connected
2023-07-21 15:54:06.968 27874-28498 TorWrapper              org.onionshare.android.nightly       I  NOTICE Bridge 'iguanasuicide' has both an IPv4 and an IPv6 address.  Will prefer using its IPv4 address (92.243.15.235:9001) based on the configured Bridge address.
2023-07-21 15:54:06.969 27874-28498 TorWrapper              org.onionshare.android.nightly       I  NOTICE new bridge descriptor 'iguanasuicide' (fresh): $477EAD3C04036B48235F1F27FC91420A286A4B7F~iguanasuicide [5r7wpZz59M7J8ku01gLktH3z8ppnmetXGRaA+Abwnzs] at 92.243.15.235 and [2001:4b98:dc0:41:216:3eff:fefd:ceca]
2023-07-21 15:54:07.069 27874-28498 TorWrapper              org.onionshare.android.nightly       I  NOTICE Bridge 'Unnamed' has both an IPv4 and an IPv6 address.  Will prefer using its IPv4 address (213.108.108.145:17674) based on the configured Bridge address.
2023-07-21 15:54:07.070 27874-28498 TorWrapper              org.onionshare.android.nightly       I  NOTICE new bridge descriptor 'Unnamed' (fresh): $A39C0FE47963B6E8CFE9815549864DE544935A31~Unnamed [+/f2KNohpQt5EW/lufYYRYXdOl9Y2LrUPczVFnuRw0Q] at 213.108.108.145 and [2a00:c6c0:0:119:3::1]
2023-07-21 15:54:07.071 27874-28498 TorWrapper              org.onionshare.android.nightly       I  OR connection CONNECTED
2023-07-21 15:54:07.071 27874-28498 TorWrapper              org.onionshare.android.nightly       I  10 OR connections connected
2023-07-21 15:54:07.459 27874-28498 TorWrapper              org.onionshare.android.nightly       I  NOTICE new bridge descriptor 'Unnamed' (fresh): $D1B6871457A0AA01BF46EEB1989D89086E5E6825~Unnamed [LEg2u4VDg7onqM6V9PLIzqi0psO9ncZWx9ZdSso3EDg] at 80.67.179.20
2023-07-21 15:54:07.460 27874-28498 TorWrapper              org.onionshare.android.nightly       I  NOTICE Bridge 'ElectricFuneral' has both an IPv4 and an IPv6 address.  Will prefer using its IPv4 address (195.201.202.125:443) based on the configured Bridge address.
2023-07-21 15:54:07.460 27874-28498 TorWrapper              org.onionshare.android.nightly       I  NOTICE new bridge descriptor 'ElectricFuneral' (fresh): $D44C0BC5AF900547704BCE5062E4B169672120E8~ElectricFuneral [bhTGOZ0CbbkNi5fSwtg484j5xmJVmS21J34SGd462oc] at 195.201.202.125 and [2a01:4f8:13a:2064::106]
2023-07-21 15:54:07.460 27874-28498 TorWrapper              org.onionshare.android.nightly       I  NOTICE new bridge descriptor 'ibksturm02' (fresh): $05E222E2A8C234234FE0CEB58B08A93B8FC360DB~ibksturm02 [W5ReYLo3OUKXpa3FMhbCY0L+P4lxXaABV0P/xLXAUyI] at 213.196.191.96
2023-07-21 15:54:07.460 27874-28498 TorWrapper              org.onionshare.android.nightly       I  NOTICE Bridge 'conceptbr1' has both an IPv4 and an IPv6 address.  Will prefer using its IPv4 address (65.21.240.163:33245) based on the configured Bridge address.
2023-07-21 15:54:07.460 27874-28498 TorWrapper              org.onionshare.android.nightly       I  NOTICE new bridge descriptor 'conceptbr1' (fresh): $20BD59649212CFE7412BFC9B94C3CCCFD8F807A8~conceptbr1 [grrpCZ6smF3f+kZ1lKWPEOTKeYNXNpwXrmisQDsfFgI] at 65.21.240.163 and [2a01:4f9:c010:85bc::1]
2023-07-21 15:54:07.462 27874-28498 TorWrapper              org.onionshare.android.nightly       I  OR connection CONNECTED
2023-07-21 15:54:07.462 27874-28498 TorWrapper              org.onionshare.android.nightly       I  11 OR connections connected
2023-07-21 15:54:07.462 27874-28498 TorWrapper              org.onionshare.android.nightly       I  OR connection CONNECTED
2023-07-21 15:54:07.462 27874-28498 TorWrapper              org.onionshare.android.nightly       I  12 OR connections connected
2023-07-21 15:54:07.462 27874-28498 TorWrapper              org.onionshare.android.nightly       I  OR connection CONNECTED
2023-07-21 15:54:07.462 27874-28498 TorWrapper              org.onionshare.android.nightly       I  13 OR connections connected
2023-07-21 15:54:07.462 27874-28498 TorWrapper              org.onionshare.android.nightly       I  OR connection CONNECTED
2023-07-21 15:54:07.462 27874-28498 TorWrapper              org.onionshare.android.nightly       I  14 OR connections connected
2023-07-21 15:54:07.502 27874-28498 TorWrapper              org.onionshare.android.nightly       I  OR connection CONNECTED
2023-07-21 15:54:07.502 27874-28498 TorWrapper              org.onionshare.android.nightly       I  15 OR connections connected
2023-07-21 15:54:07.534 27874-28498 TorWrapper              org.onionshare.android.nightly       I  OR connection CONNECTED
2023-07-21 15:54:07.534 27874-28498 TorWrapper              org.onionshare.android.nightly       I  16 OR connections connected
2023-07-21 15:54:07.662 27874-28498 TorWrapper              org.onionshare.android.nightly       I  NOTICE Bridge 'kallio' has both an IPv4 and an IPv6 address.  Will prefer using its IPv4 address (87.100.193.2:9010) based on the configured Bridge address.
2023-07-21 15:54:07.662 27874-28498 TorWrapper              org.onionshare.android.nightly       I  NOTICE new bridge descriptor 'kallio' (fresh): $13FB63452AADFA082BD2BC3E1E320AD301F07877~kallio [nzUXE7EOdZyPm4Xj5YsCh7gl2x/uFlo8blWtlR6FnTw] at 87.100.193.2 and [2001:14ba:ab00:bb48::1]
2023-07-21 15:54:07.852 27874-28498 TorWrapper              org.onionshare.android.nightly       I  OR connection CONNECTED
2023-07-21 15:54:07.852 27874-28498 TorWrapper              org.onionshare.android.nightly       I  17 OR connections connected
2023-07-21 15:54:07.853 27874-28498 TorWrapper              org.onionshare.android.nightly       I  NOTICE new bridge descriptor 'KauBridgePale' (fresh): $2D82C2E354D531A68469ADF7F878FA6060C6BACA~KauBridgePale [oggxcozpXaXS3PpTGVlr9kolycQ7wXQx0f67YECNprA] at 193.11.166.194
2023-07-21 15:54:07.853 27874-28498 TorWrapper              org.onionshare.android.nightly       I  NOTICE new bridge descriptor 'KauBridgeBlue' (fresh): $86AC7B8D430DAC4117E9F42C9EAED18133863AAF~KauBridgeBlue [WtmZ2uv4Y7guCnLU+hI0WbjGQlZMkREbsLZltXaXadQ] at 193.11.166.194
2023-07-21 15:54:08.138  4801-4801  audit                   auditd                               E  type=1300 audit(1689951248.132:5375): arch=c00000b7 syscall=29 success=no exit=-13 a0=11 a1=894b a2=784362f0b4 a3=7fff28d1e8 items=0 ppid=27874 pid=28494 auid=4294967295 uid=10003 gid=10003 euid=10003 suid=10003 fsuid=10003 egid=10003 sgid=10003 fsgid=10003 tty=(none) ses=4294967295 comm="libtor.so" exe="/data/app/~~K0sq_Zw3LvcFPaZR5qjM8g==/org.onionshare.android.nightly-Y4X77aileIM8AKNX_5gBRw==/lib/arm64/libtor.so" subj=u:r:untrusted_app:s0:c3,c256,c512,c768 key=(null)
2023-07-21 15:54:08.138 27874-28498 TorWrapper              org.onionshare.android.nightly       I  NOTICE new bridge descriptor 'KauBridgeDot' (fresh): $1AE2C08904527FEA90C4C4F8C1083EA59FBC6FAF~KauBridgeDot [r7OXmH6vfpl+GJqqfajI6Zap9dSNbxbdisopTdyS6QE] at 193.11.166.194
2023-07-21 15:54:08.139 27874-28498 TorWrapper              org.onionshare.android.nightly       I  NOTICE new bridge descriptor 'zipfelmuetze' (fresh): $91A6354697E6B02A386312F68D82CF86824D3606~zipfelmuetze [7YRAoSWJ+wYD5j59olKF8/WmZ1yKaG6hoIZuOUFvsKk] at 85.31.186.26
2023-07-21 15:54:08.140 27874-28498 TorWrapper              org.onionshare.android.nightly       I  NOTICE new bridge descriptor 'Lisbeth' (fresh): $CDF2E852BF539B82BD10E27E9115A31734E378C2~Lisbeth [DTpHLkU/hQso35JMGpJOY/6aWYjvfaup5LFU0+2WBII] at 192.95.36.142
2023-07-21 15:54:08.310 27874-28498 TorWrapper              org.onionshare.android.nightly       I  NOTICE new bridge descriptor 'griinchux' (fresh): $011F2599C0E9B27EE74B353155E244813763C3E5~griinchux [Gf6Zetna0onvhOv8OLzeiUUfjv5zV6DFl82+x8wEt7A] at 85.31.186.98
2023-07-21 15:54:08.593 27874-28498 TorWrapper              org.onionshare.android.nightly       I  NOTICE new bridge descriptor 'PraxedisGuerrero' (fresh): $5EDAC3B810E12B01F6FD8050D2FD3E277B289A08~PraxedisGuerrero [qSueJICMYZBsv2+lFaemfzHkPZxDER2ZDeQdiLDMIDY] at 51.222.13.177
2023-07-21 15:54:09.353 27874-28498 TorWrapper              org.onionshare.android.nightly       I  OR connection CONNECTED
2023-07-21 15:54:09.354 27874-28498 TorWrapper              org.onionshare.android.nightly       I  18 OR connections connected
2023-07-21 15:54:10.876 27874-28498 TorWrapper              org.onionshare.android.nightly       I  NOTICE new bridge descriptor 'hopperlab' (fresh): $10A6CD36A537FCE513A322361547444B393989F0~hopperlab [xpuRzwsil54EG1HASFMlYoBCpMVFwYthoul2fMZPzQM] at 146.57.248.225
2023-07-21 15:54:10.877 27874-28498 TorWrapper              org.onionshare.android.nightly       I  NOTICE We now have enough directory information to build circuits.
2023-07-21 15:54:11.772 27874-28498 TorWrapper              org.onionshare.android.nightly       I  OR connection FAILED
2023-07-21 15:54:11.813 27874-28498 TorWrapper              org.onionshare.android.nightly       I  OR connection FAILED
2023-07-21 15:54:14.873 27874-28498 TorWrapper              org.onionshare.android.nightly       I  OR connection LAUNCHED
2023-07-21 15:54:15.869 27874-28498 TorWrapper              org.onionshare.android.nightly       I  OR connection LAUNCHED
2023-07-21 15:54:17.327 27874-28498 TorWrapper              org.onionshare.android.nightly       I  OR connection CONNECTED
2023-07-21 15:54:17.327 27874-28498 TorWrapper              org.onionshare.android.nightly       I  19 OR connections connected
2023-07-21 15:54:18.045 27874-28498 TorWrapper              org.onionshare.android.nightly       I  NOTICE new bridge descriptor 'GeorgetownPontem' (fresh): $74FAD13168806246602538555B5521A0383A1875~GeorgetownPontem [ZWCs12Nk8NW7PEGQ2r5ZBF6PLXFnT+o5W3TS5QC/JGM] at 209.148.46.65
2023-07-21 15:54:21.025 27874-28498 TorWrapper              org.onionshare.android.nightly       I  OR connection FAILED
2023-07-21 15:54:22.707 27874-28498 TorWrapper              org.onionshare.android.nightly       I  OR connection LAUNCHED
2023-07-21 15:54:27.888 27874-28498 TorWrapper              org.onionshare.android.nightly       I  OR connection FAILED
2023-07-21 15:54:28.820 27874-28498 TorWrapper              org.onionshare.android.nightly       I  OR connection LAUNCHED
2023-07-21 15:54:34.336 27874-28498 TorWrapper              org.onionshare.android.nightly       I  OR connection FAILED
2023-07-21 15:54:36.841 27874-28498 TorWrapper              org.onionshare.android.nightly       I  OR connection LAUNCHED
2023-07-21 15:54:38.114 27874-28498 TorWrapper              org.onionshare.android.nightly       I  OR connection FAILED
2023-07-21 15:54:38.117 27874-28498 TorWrapper              org.onionshare.android.nightly       I  WARN Proxy Client: unable to connect OR connection (handshaking (proxy)) with 45.145.95.6:27015 ID=<none> RSA_ID=C5B7CD6946FF10C5B3E89691A7D3F2C122D2117C ("general SOCKS server failure")
2023-07-21 15:54:41.914 27874-28498 TorWrapper              org.onionshare.android.nightly       I  OR connection FAILED
2023-07-21 15:54:42.033 27874-28498 TorWrapper              org.onionshare.android.nightly       I  OR connection LAUNCHED
2023-07-21 15:54:50.970 27874-28498 TorWrapper              org.onionshare.android.nightly       I  OR connection LAUNCHED
2023-07-21 15:54:53.978 27874-28498 TorWrapper              org.onionshare.android.nightly       I  OR connection LAUNCHED
2023-07-21 15:54:58.708 27874-28498 TorWrapper              org.onionshare.android.nightly       I  WARN Proxy Client: unable to connect OR connection (handshaking (proxy)) with 45.145.95.6:27015 ID=<none> RSA_ID=C5B7CD6946FF10C5B3E89691A7D3F2C122D2117C ("general SOCKS server failure")
2023-07-21 15:55:00.366 27874-28028 RenewableWakeLock       org.onionshare.android.nightly       I  Renewing wake lock org.onionshare.android.nightly
2023-07-21 15:55:09.069 27874-28029 o.o.a.f.FileManager     org.onionshare.android.nightly       I  Got cancelled, deleting zip file...
2023-07-21 15:55:09.308 27874-28029 o.o.a.f.FileManager     org.onionshare.android.nightly       E  Error zipping files: kotlinx.coroutines.JobCancellationException: ScopeCoroutine was cancelled
2023-07-21 15:55:09.309 27874-28029 o.o.a.ShareManager      org.onionshare.android.nightly       I  Start job to finished.
2023-07-21 15:55:09.309 27874-28029 o.o.a.t.TorManager      org.onionshare.android.nightly       I  Stopping...
2023-07-21 15:55:09.310 27874-27874 o.o.a.t.TorManager      org.onionshare.android.nightly       I  new state: STOPPING
2023-07-21 15:55:09.378 27874-28029 RenewableWakeLock       org.onionshare.android.nightly       I  Releasing wake lock org.onionshare.android.nightly
2023-07-21 15:55:09.378 27874-27874 o.o.a.t.TorManager      org.onionshare.android.nightly       I  new state: STOPPED
2023-07-21 15:55:09.379 27874-28019 TorWrapper              org.onionshare.android.nightly       I  Tor exited with value 0
2023-07-21 15:55:09.386 27874-27874 o.o.a.t.ShareService    org.onionshare.android.nightly       D  onDestroy
```

A few things to note here:
* The log starts when the "start sharing" button is pressed
* Starting the Tor process takes a couple of seconds
* TorManager logs the state changes received from onionwrapper (which shows that they're being observed)
* TorManager starts the check job that calls `checkStartupProgress()`
* ShareManager's Tor task returns (as expected) and it starts the observer task
* After a couple more seconds, TorManager logs the onionwrapper state `CONNECTED`, meaning that Tor has successfully bootstrapped
* ShareManager logs "New share state: Starting(zipPercent=0, torPercent=95)"
* Nothing happens for two minutes (except the wake lock being renewed)
* TorManager logs "Tor startup is not making progress" (but Tor has already started)
* TorManager logs "Getting bridges from Moat"
* Seemingly Moat returns an empty list of bridges, because we don't see "Error getting bridges from moat" or "Using bridges from Moat"
* Instead, 3 seconds after "Getting bridges from Moat" we see "Using built-in bridges"
* After applying the bridge settings, TorManager logs "Waiting for Tor to start" again, as expected
* But then it immediately logs "Tor startup is not making progress" again, without actually waiting
* At this point TorManager think it's out of options for starting Tor, so it logs "Could not connect to Tor" and sets the TorState to `FailedToConnect`
* ShareManager cancels the start job and waits for it to finish
* Meanwhile Tor is trying to bootstrap via the built-in bridges, because we haven't stopped it
* Finally, a minute after cancellation, the zip task finishes zipping the file and throws `JobCancellationException`
* ShareManager's start job has finished, so now it stops Tor
* We end up with Tor stopped, zip file deleted, and error state shown in the UI - all of which would be good if Tor had actually failed to start, but it succeeded

Reply:
I think the issue here is that we don't upload a HS_DESC for some reason. I've seen this only once, but apparently, it happens sometimes. In `TorManager`, we don't consider things as started before uploading the descriptor, but maybe we should differentiate here. Still, if we *never* upload a descriptor, not sure how this helps.

Could this be the issue that Tor somehow already had it uploaded and it is "fresh" enough not needing re-upload?

> Seemingly Moat returns an empty list of bridges, because we don't see "Error getting bridges from moat" or "Using bridges from Moat"

Should we log when the bridges were empty?

> Instead, 3 seconds after "Getting bridges from Moat" we see "Using built-in bridges"

That's actually the currently expected behavior if moat didn't give us bridges.

> But then it immediately logs "Tor startup is not making progress" again, without actually waiting

This is probably because of `if (now - s.lastProgressTime > TOR_START_TIMEOUT_SINCE_LAST_PROGRESS)`, so no matter how much `now` advances, this condition will always be true, if it was true once. Not sure how to handle this. Remove the check? Reset `lastProgressTime` when starting to wait?

> Meanwhile Tor is trying to bootstrap via the built-in bridges, because we haven't stopped it

This doesn't sound right, but stopping Tor before the start job was cancelled also doesn't sound right. Maybe if we implemented cancelling ongoing file copy operations (should be quick), this wouldn't be an issue as cancelling the start job would be faster and we'd move on to stopping Tor?

Reply:
> Could this be the issue that Tor somehow already had it uploaded and it is "fresh" enough not needing re-upload?

Maybe, but I think it's suspicious that it reliably happens on this branch, and reliably doesn't happen on the main branch. I'll try a couple more times to see if it's really reliable.

> Should we log when the bridges were empty?

Sure, might make the logs a bit clearer.

> That's actually the currently expected behavior if moat didn't give us bridges.

Understood, I just wanted to report all the main events from the log.

>  Not sure how to handle this. Remove the check? Reset lastProgressTime when starting to wait?

Resetting it sounds good.

> This doesn't sound right, but stopping Tor before the start job was cancelled also doesn't sound right. Maybe if we implemented cancelling ongoing file copy operations (should be quick), this wouldn't be an issue as cancelling the start job would be faster and we'd move on to stopping Tor?

I think there are two entangled issues here:
1. We think Tor has failed to bootstrap, but actually there are still methods it could try. This should be fixed by resetting `lastProgressTime` as you suggested.
2. Tor really has failed to bootstrap within the allowed time, but then it continues trying to bootstrap while we wait for the files to be zipped. In my view this isn't a major problem. It would be nice and tidy if we disabled Tor's network connection after all bootstrap methods failed, but it's not important. The more important thing is to interrupt the zipping somehow so that this situation is short-lived and we can give the user the cancellation they asked for.

Reply:
> Could this be the issue that Tor somehow already had it uploaded and it is "fresh" enough not needing re-upload?

On second thoughts this shouldn't be the case because we're using a new HS each time.

The reason for not seeing HS publication on this branch for the scenario above, but seeing it on main, is that on this branch we don't call publishOnionService() until the zip task completes. So it's not a problem.

Reply:
ok I added two commits: one resets the progress time and the other cancels zipping while we copy a file. There's also a beginning for tests now. These could be extended to cover more components.